compiler-core/src/javascript/expression.rs

```diff
@@ -222,57 +222,112 @@ impl<'module> Generator<'module> {
         // Collect all the values used in segments.
         let segments_array = array(segments.iter().map(|segment| {
             let value = self.not_in_tail_position(|gen| gen.wrap_expression(&segment.value))?;
-            match segment.options.as_slice() {
-                // Ints
-                [] | [Opt::Int { .. }] => Ok(value),
 
-                // Sized ints
-                [Opt::Size { value: size, .. }] => {
-                    let size_int = match *size.clone() {
-                        TypedExpr::Int {
-                            location: _,
-                            typ: _,
-                            value,
-                        } => value.parse().unwrap_or(0),
-                        _ => 0,
-                    };
-                    if size_int > 0 && size_int % 8 != 0 {
-                        return Err(Error::Unsupported {
-                            feature: "Non byte aligned array".into(),
-                            location: segment.location,
-                        });
-                    }
-                    self.tracker.sized_integer_segment_used = true;
-                    let size = self.not_in_tail_position(|gen| gen.wrap_expression(size))?;
-                    Ok(docvec!["sizedInt(", value, ", ", size, ")"])
+            if segment.type_ == crate::type_::int() || segment.type_ == crate::type_::float() {
```

Comment:
This whole function has become hard to read and understand, and I'm not confident I can review it well enough to catch any bugs with the design here.

Could we instead split into a parsing step which turns it into a struct or an enum, and then use that to perform code generation. Thank you